### PR TITLE
v3: link with as child pattern

### PR DESCRIPTION
**Pull request by jeetiss**
`<Link asChild />` feature

### usage

```
// right

<Link to='href' ref={ref}>
  link
</Link>

<Link to='href' ref={ref} className={'hello'}>
  link
</Link>

<Link to='href' asChild>
  <Button>hello world</Button>
</Link>

<Link to='href' asChild>
  <Button ref={ref}>hello world</Button>
</Link>

// wrong

<Link to='href' asChild ref={ref} className={'noo'}>
  <Button ref={ref}>hello world</Button>
</Link>

<Link to='href' asChild>
  text
</Link>
```

### CODE REVIEW

File: packages/wouter/src/index.js [hunk 1]
```diff
@@ -204,17 +204,29 @@ export const Link = forwardRef((props, ref) => {
...
-  // wraps children in `a` if needed
-  const extraProps = {
-    // handle nested routers and absolute paths
-    href: href[0] === "~" ? href.slice(1) : router.base + href,
+  // handle nested routers and absolute paths
+  const link = href[0] === "~" ? href.slice(1) : router.base + href;
+
+  if (asChild) {
+    if (!isValidElement(children)) {
+      throw Error("Only one child allowed");
```
Comment (jeetiss) [inline packages/wouter/src/index.js:212]: I'm not sure about `error` here. is it okay to trow error when children are invalid?

@molefrog what do you think?

File: packages/wouter-preact/types/index.d.ts [hunk 2]
```diff
@@ -81,11 +81,12 @@ export type NavigationalProps<
...
-  JSX.HTMLAttributes,
-  "href"
-> &
-  NavigationalProps<H>;
+type AsChildProps<RequiredProps, DefaultElementProps> =
+  | ({ asChild?: false } & RequiredProps & DefaultElementProps)
+  | ({ asChild: true; children: ComponentChildren } & RequiredProps);
+
+export type LinkProps<H extends BaseLocationHook = BrowserLocationHook> =
+  AsChildProps<NavigationalProps<H>, JSX.HTMLAttributes>;
```
Comment (jeetiss) [inline packages/wouter-preact/types/index.d.ts:89]: I have no experience with preact types, could you review it with more passion?

File: packages/wouter/test/link.test-d.tsx [hunk 3]
```diff
@@ -1,44 +1,60 @@
...
-      <Link href="/" state={undefined}>
-        test
-      </Link>
-    );
-    assertType(
-      <Link href="/" state="string">
-        test
-      </Link>
-    );
+  it("should throw error when `to` and `href` props are used in same time", () => {
```
Comment (molefrog) [inline packages/wouter/test/link.test-d.tsx:12]: I would rephrase this, because technically it doesn't throw an error. Maybe:
```suggestion
  it("does not allow `to` and `href` props to be used at the same time", () => {
```

File: packages/wouter/test/link.test-d.tsx [hunk 4]
```diff
@@ -1,44 +1,60 @@
...
+    // @ts-expect-error
+    <Link to="/hello" href="/world">
+      Hello
+    </Link>;
+  });
+
+  it("should inherit props from `HTMLAnchorElements`", () => {
```
Comment (molefrog) [inline packages/wouter/test/link.test-d.tsx:19]:
```suggestion
  it("should inherit props from `HTMLAnchorElement`", () => {
```

File: packages/wouter/test/link.test-d.tsx [hunk 5]
```diff
@@ -56,19 +72,61 @@ describe("Link's ref", () => {
...
-    <Link to="/">
-      <Component ref={ref}></Component>
+  it("should throw error when `to` and `href` props are used in same time", () => {
+    // @ts-expect-error
+    <Link to="/hello" href="/world" asChild>
+      <a>Hello</a>
+    </Link>;
+  });
+
+  it("should throw error when unsupported props are provided", () => {
```
Comment (molefrog) [inline packages/wouter/test/link.test-d.tsx:94]: Same here.

File: packages/wouter/test/link.test-d.tsx [hunk 6]
```diff
@@ -56,19 +72,61 @@ describe("Link's ref", () => {
...
+    // @ts-expect-error
+    <Link to="/" asChild ref={null}>
+      <a>Hello</a>
     </Link>;
   });
-});
 
-const Component = React.forwardRef<{ hello: "world" }>((props, ref) => {
-  return <>test</>;
+  it("should throw error when to and href props are used in same time", () => {
```
Comment (molefrog) [inline packages/wouter/test/link.test-d.tsx:116]: And here.

File: packages/wouter/test/link.test-d.tsx [hunk 7]
```diff
@@ -56,19 +72,61 @@ describe("Link's ref", () => {
...
   });
+});
 
-  it("should work with composed components", () => {
-    const ref = React.useRef<React.ElementRef<typeof Component>>(null);
+describe("<Link /> in asChild mode", () => {
+  it("should work", () => {
+    <Link to="/" asChild>
+      <a>Hello</a>
+    </Link>;
```
Comment (molefrog) [inline packages/wouter/test/link.test-d.tsx:84]: I would also test custom component just in case to show that not only links are allowed here.

File: packages/wouter/test/link.test-d.tsx [hunk 8]
```diff
@@ -1,44 +1,60 @@
...
+  });
+
+  it("should support other navigation params", () => {
+    <Link href="/" state={{ a: "foo" }}>
+      test
+    </Link>;
+
+    <Link href="/" replace>
+      test
+    </Link>;
```
Comment (molefrog) [inline packages/wouter/test/link.test-d.tsx:44]: At the same time, using anything other than boolean as `replace` should not be allowed. Could you test that as well?

File: packages/wouter/test/link.test-d.tsx [hunk 9]
```diff
@@ -1,44 +1,60 @@
...
+
+    <Link href="/" state={undefined}>
+      test
+    </Link>;
   });
```
Comment (molefrog) [inline packages/wouter/test/link.test-d.tsx:49]: Imo, we should also test generic Link type with custom navigation hook. Let's say for example `useNetworkLocation` and `navigate({ host: "whatever", retries: 4 })`.

File: packages/wouter/test/link.test.tsx [hunk 10]
```diff
@@ -1,206 +1,221 @@
...
+  });
 });
 
-it("renders `href` with basepath", () => {
-  const { getByTestId } = render(
-    <Router base="/app">
-      <Link href="/dashboard" data-testid="link" />
-    </Router>
-  );
+describe("<Link /> in asChild mode", () => {
```
Comment (molefrog) [inline packages/wouter/test/link.test.tsx:169]:
```suggestion
describe("<Link /> with `asChild` props", () => {
```

File: packages/wouter/test/link.test.tsx [hunk 11]
```diff
@@ -1,206 +1,221 @@
...
+  it("works for any other elements as well", () => {
```
Comment (molefrog) [inline packages/wouter/test/link.test.tsx:170]: I think the test case label is a bit confusing. How about:

```suggestion
  it("when `asChild` is not specified, wraps the children in an <a />", () => {
```

File: packages/wouter/types/index.d.ts [hunk 12]
```diff
@@ -113,8 +107,18 @@ export function Redirect<H extends BaseLocationHook = BrowserLocationHook>(
   context?: any
 ): ReactElement<any, any> | null;
 
+type AsChildProps<RequiredProps, DefaultElementProps> =
```
Comment (molefrog) [inline packages/wouter/types/index.d.ts:110]: What do you think about renaming `RequiredProps` to something like `NavigateOptions`?

**Review by molefrog — CHANGES_REQUESTED** [review]

File: packages/wouter-preact/types/index.d.ts (hunk 2 above)
Comment (molefrog) [reply packages/wouter-preact/types/index.d.ts:89]: I am not an expert either but could you try `JSX.IntrinsicElements["a"]`? This should give attributes tailored to `a` (mostly to get the proper type in event handlers).

File: packages/wouter/types/index.d.ts (hunk 12 above)
Comment (jeetiss) [reply packages/wouter/types/index.d.ts]: This is a generic type that can be used with other components, not only with links.

I renamed this prop to `ComponentProps`, is this a better name?

File: packages/wouter/test/link.test-d.tsx (hunk 7 above)
Comment (jeetiss) [reply packages/wouter/test/link.test-d.tsx:107]: typescript can't check children types, so if a regular one works, custom would work too

File: packages/wouter-preact/types/index.d.ts (hunk 2 above)
Comment (jeetiss) [reply packages/wouter-preact/types/index.d.ts]: `JSX.IntrinsicElements["a"]` nope
it doesn't work

**Review by molefrog — APPROVED** [review]
👍

File: packages/wouter-preact/types/index.d.ts (hunk 2 above)
Comment (molefrog) [reply packages/wouter-preact/types/index.d.ts]: Ok!

**Review by molefrog — APPROVED** [review]
